src/Admin/models/useStorage.js

```diff
@@ -0,0 +1,39 @@
+import { useState, useEffect } from "react";
+import firebase from "../../common/lib/firebase";
+
+export const useStorage = (file) => {
+  const [progress, setProgress] = useState(null);
+  const [error, setError] = useState(null);
+  const [url, setUrl] = useState(null);
+
+  useEffect(() => {
+    if (!file) return;
+    console.log(typeof file);
+
+    console.log("Storage Hook Fired");
+    console.log(file.name);
+
+    const projectStorage = firebase.storage();
+
+    const storageRef = projectStorage.ref(file.name); // Need to change to id
+
+    storageRef.put(file).on(
+      "state_changed",
+      (snap) => {
+        let percentage = (snap.bytesTransferred / snap.totalBytes) * 100;
+        setProgress(percentage);
+      },
+      (err) => {
+        setError(err);
+      },
+      async () => {
+        const url = await storageRef.getDownloadURL();
+        setUrl(url);
+        setProgress(null);
+      }
+    );
+  }, [file]);
```

Comment:
If we think about this, it seems we want to perform some procedure from outside this hook.  The key here is we just want it to be able to manage the state during that procedure.  My suggestion would be to, instead of `useEffect`, create a function which takes `file` as a parameter.  Then return this function as part of what you return with `progress, url, error`.

However, it might be worth considering doing all of this in the submit functionality (nothing needs to be uploaded until the user "confirms" it by submitting).

Reply:
Let me have a play around with this and see how I go